### PR TITLE
`constrained-generators` cleanup for hackage

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

This is also in preparation for moving the specs for various STS rules out of `cardano-ledger-test`. @lehins, @Soupstraw, and @TimSheard what are your wishes for this? I have the following TODO that I can recall:

- [ ] Type class for specs used in STS rules (@Soupstraw did you end up doing that in the last PR or not?)
- [ ] Move the STS specs out of `cardano-ledger-test`
- [x] Rename the `Spec` type so that it doesn't clash with `HSpec.Spec`
- [x] Migrate away from `tasty` to `hspec`

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by lehins** [comment]
> I have the following TODO that I can recall:

I don't think I can add anything else to the TODO list at the moment. The plan sounds good to me.

**Comment by MaximilianAlgehed** [comment]
In order to focus on fixing a bug for @TimSheard I've decided to push the top two TODOs for the next PR. Could we get this reviewed and merged ASAP?

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -2326,17 +2332,17 @@ genNumList ::
...
   normFoldS <- normalize foldSIn
   explain
     [ "Can't generate list of ints with fold constraint"
-    , show $ "  elemSpec =" <+> pretty elemSIn
-    , show $ "  normElemSpec =" <+> pretty normElemS
-    , show $ "  foldSpec =" <+> pretty foldSIn
+    , "  elemSpec = " ++ show elemSIn
+    , "  normElemSpec = " ++ show normElemS
+    , "  foldSpec = " ++ show foldSIn
     ]
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Base.hs:2397]: Not a good idea in my opinion to change Pretty to Show.

**Review by TimSheard — CHANGES_REQUESTED** [review]
There seem to be 4 kinds of changes in this PR
1) Changing the type Spec to Specification
2) Adding a bit of new functionality to Conway.Instances
3) Moving from Pretty to Show
4) Splitting the tests into multiple modules. One for List, Set, Map ,Property etc.

Vast majority of changes (> 90%) involve the Spec to Specification

In my opinion moving from Pretty to Show is ill-advised, and should be reverted.

File: libs/constrained-generators/src/Constrained/Base.hs (hunk 1 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:2397]: This is just a refactoring:
```haskell
instance HasSpec fn a => Show (Specification fn a) where
  showsPrec d = shows . pretty . WithPrec d
```

**Review by lehins — APPROVED** [review]
Looks awesome!

Here is a side comment to what @TimSheard said:

> In my opinion moving from Pretty to Show is ill-advised, and should be reverted.

I am totally in favor of **not** using the custom Pretty instances. Pretty has huge downsides for the purposes of testing and debugging:
* it is very confusing, because the output is free form and without looking at the actual instance it is hard to infer the meaning of the output or even the type that is being printed.
* it has huge maintenance overhead. Every time any of the ledger types changes we need to go and manually adjust the instance.
* a bug in Pretty instance could lead to either a buggy test or at the very least hard time during debugging of an actual issue.
* Pretty instances themselves are untested and even right now bunch of the instances are incomplete.

My goal is one day to completely get rid of Pretty interface in the ledger codebase.

**Comment by TimSheard** [comment]
We are not paying attention here.  The only Pretty Instances I want to keep are the ones for 
Term, Spec, Pred, etc.  These are the types native to the this package alone. These things can get large
and when a Spec fails, being able to understand what is wrong is helped by the pretty printing. I agree with Alexey, we don't need any Pretty instance for Ledger types.

**Comment by MaximilianAlgehed** [comment]
I agree on the ledger types. For the constraint types you do need pretty instances that make things readable (you don't want to read abstract syntax trees!)

Anyway, @TimSheard could you approve the PR? The issue you were concerned about is moot because the show instance for Spec is just the pretty instance.

**Review by TimSheard — APPROVED** [review]
OK I approve.

**Pull request merged**